sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -371,6 +384,18 @@ case class Invoke(
     }
 
   private lazy val encodedFunctionName = ScalaReflection.encodeFieldNameToIdentifier(functionName)
+  private var isDeterministic: Boolean = true
```

Comment:
Note that having a `var` here breaks the convention that case classes are immutable, and also breaks `copy` and equality checks which would ignore this value.

Reply:
Hmm I'm not sure whether there's a better way. I tried to make the parameter non-default but then it seems will break the API compatibility.

Reply:
I think it should be possible using secondary constructor in the companion object. Let me take a look.

Reply:
My suggestion is to keep `isDeterministic: boolean = true` in the constructor, and add `def this` and and `apply` in a companion object:

```
case class StaticInvoke(
  staticObject: Class[_],
  dataType: DataType,
  functionName: String,
  arguments: Seq[Expression] = Nil,
  inputTypes: Seq[AbstractDataType] = Nil,
  propagateNull: Boolean = true,
  returnNullable: Boolean = true,
  isDeterminitic: Boolean = false) {

  def this(
	staticObject: Class[_],
	dataType: DataType,
	functionName: String,
	arguments: Seq[Expression],
	inputTypes: Seq[AbstractDataType],
	propagateNull: Boolean,
	returnNullable: Boolean
  ) = {
  this(staticObject, dataType, functionName, arguments, inputTypes, propagateNull, returnNullable, true)
  }

}

object StaticInvoke {
  def apply(
	staticObject: Class[_],
	dataType: DataType,
	functionName: String,
	arguments: Seq[Expression],
	inputTypes: Seq[AbstractDataType],
	propagateNull: Boolean,
	returnNullable: Boolean
  ): StaticInvoke = StaticInvoke(staticObject, dataType, functionName, arguments, inputTypes, propagateNull, returnNullable, true)
}
```

The two methods added without default parameters represent what existing compiled libraries are linked with. I tested this suggestion in a test project where I added empty traits for `DataType`, `Expression`, etc. The above suggestion takes care of the constructor compatibility, but leaves a few unavoidable binary compatibility issues that should be inconsequential (pertaining to `tupled`, `curried` and `copy` method):

```
[error] compat: Failed binary compatibility check against default:compat_2.13:0.1.0! Found 4 potential problems (filtered 2)
[error]  * static method tupled()scala.Function1 in class x.StaticInvoke does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("x.StaticInvoke.tupled")
[error]  * static method curried()scala.Function1 in class x.StaticInvoke does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("x.StaticInvoke.curried")
[error]  * method copy(java.lang.Class,x.DataType,java.lang.String,scala.collection.immutable.Seq,scala.collection.immutable.Seq,Boolean,Boolean)x.StaticInvoke in class x.StaticInvoke does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("x.StaticInvoke.copy")
[error]  * the type hierarchy of object x.StaticInvoke is different in current version. Missing types {scala.runtime.AbstractFunction7}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("x.StaticInvoke$")
```

Reply:
+1 to @thesamet 's suggestion

We can also consider cleaning up the legacy/compatibility versions of `this`/`apply` in the next minor release and keeping them around only for patch releases?

Reply:
It would make sense to add a comment to have `apply` and `this` removed for Spark 3.3.0

Reply:
Ah cool, this is neat. Let me switch to use this.

Reply:
Hmm what's the benefit of removing these before the Spark 3.3.0 release? won't that cause the compatibility issue again? say frameless wants to work with both Spark 3.2.x and 3.3.x

Reply:
The intent of the suggestion was to do this on or after the 3.3.0 release. It's ok to defer, but binary incompatibilities between 3.2.x and 3.3.0 are more expected. This is just to balance how long you want to carry the noise in the code caused by the additional constructors.